src/Psl/Type/Internal/NonEmptyVecType.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use Psl;
+use Psl\Str;
+use Psl\Type;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+
+use function is_iterable;
+
+/**
+ * @template Tv
+ *
+ * @extends Type\Type<non-empty-list<Tv>>
+ *
+ * @internal
+ */
+final class NonEmptyVecType extends Type\Type
+{
+    /**
+     * @param Type\TypeInterface<Tv> $value_type
+     *
+     * @throws Psl\Exception\InvariantViolationException If $value_type is optional.
+     */
+    public function __construct(
+        private Type\TypeInterface $value_type
+    ) {
+        Psl\invariant(!$value_type->isOptional(), 'Optional type must be the outermost.');
+    }
+
+    /**
+     * @psalm-assert-if-true non-empty-list<Tv> $value
+     */
+    public function matches(mixed $value): bool
+    {
+        if (!is_array($value)) {
+            return false;
+        }
+
+        if ([] === $value) {
+            return false;
+        }
+
+        $index = 0;
+        foreach ($value as $k => $v) {
+            if ($index !== $k) {
+                return false;
+            }
+
+            if (!$this->value_type->matches($v)) {
+                return false;
+            }
+
+            $index++;
+        }
+
+        return true;
+    }
+
+    /**
+     * @throws CoercionException
+     *
+     * @return non-empty-list<Tv>
+     */
+    public function coerce(mixed $value): iterable
+    {
+        if (is_iterable($value)) {
+            $value_trace = $this->getTrace()
+                ->withFrame(Str\format('list<%s>', $this->value_type->toString()));
```

Comment:
```suggestion
                ->withFrame(Str\format('non-empty-vec<%s>', $this->value_type->toString()));
```

( if the vec type uses list, it should be updated to vec as well ).